### PR TITLE
BUG: interchange bitmasks not supported in interchange/from_dataframe.py

### LINKED ISSUE #49888: BUG: interchange bitmasks not supported in interchange/from_dataframe.py
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

Sorry, this is not easily reproducible as the dataframe interchange protocol for pyarrow is still work in progress but I think the error is quite clear:

```python
import pyarrow as pa
table = pa.table({"a": [1, 2, 3, None]})

exchange_df = table.__dataframe__()

from pandas.core.interchange.from_dataframe import from_dataframe
from_dataframe(exchange_df)
```

```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 53, in from_dataframe
    return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 74, in _from_dataframe
    pandas_df = protocol_df_chunk_to_pandas(chunk)
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 122, in protocol_df_chunk_to_pandas
    columns[name], buf = primitive_column_to_ndarray(col)
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 160, in primitive_column_to_ndarray
    data = set_nulls(data, col, buffers["validity"])
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 504, in set_nulls
    null_pos = buffer_to_ndarray(valid_buff, valid_dtype, col.offset, col.size)
  File "/Users/alenkafrim/repos/pyarrow-dev-9/lib/python3.9/site-packages/pandas/core/interchange/from_dataframe.py", line 395, in buffer_to_ndarray
    raise NotImplementedError(f"Conversion for {dtype} is not yet supported.")
NotImplementedError: Conversion for (<DtypeKind.BOOL: 20>, 1, 'b', '=') is not yet supported.
```

### Issue Description

I am currently working on implementing a dataframe interchange protocol for `pyarrow.Table` in Apache Arrow project (https://github.com/apache/arrow/pull/14613).

I am using pandas implementation to test that the produced `__dataframe__` object can be correctly consumed.

When consuming a `pyarrow.Table` with missing values I get an `NotImplementedError`. The bitmasks, used by PyArrow to represent nulls in a given column, can not be converted.

But if I look at the code in [from_dataframe.py](https://github.com/pandas-dev/pandas/blob/main/pandas/core/interchange/from_dataframe.py):
https://github.com/pandas-dev/pandas/blob/70121c75a0e2a42e31746b6c205c7bb9e4b9b930/pandas/core/interchange/from_dataframe.py#L405-L415

I would think this is not intentional and that the `_NP_DTYPES` should include `{1: bool}`
https://github.com/pandas-dev/pandas/blob/70121c75a0e2a42e31746b6c205c7bb9e4b9b930/pandas/core/interchange/from_dataframe.py#L393

https://github.com/pandas-dev/pandas/blob/70121c75a0e2a42e31746b6c205c7bb9e4b9b930/pandas/core/interchange/from_dataframe.py#L23-L28

### Expected Behavior

The bitmask can be converted to `ndarray` by the current pandas implementation of the dataframe interchange protocol and the code below could work for missing values also:

```python
>>> import pyarrow as pa
>>> table = pa.table({"a": [1, 2, 3, 4]})

>>> exchange_df = table.__dataframe__()
>>> exchange_df._df
pyarrow.Table
a: int64
----
a: [[1,2,3,4]]

>>> from pandas.core.interchange.from_dataframe import from_dataframe
>>> from_dataframe(exchange_df)
   a
0  1
1  2
2  3
3  4
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 87cfe4e38bafe7300a6003a1d18bd80f3f77c763
python           : 3.9.14.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.6.0
Version          : Darwin Kernel Version 21.6.0: Thu Sep 29 20:13:46 PDT 2022; root:xnu-8020.240.7~1/RELEASE_ARM64_T8101
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.5.0
numpy            : 1.22.3
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : 0.29.28
pytest           : 7.1.3
hypothesis       : 6.39.4
sphinx           : 4.3.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 8.1.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 2022.02.0
gcsfs            : 2022.02.0
matplotlib       : 3.6.2
numba            : 0.56.4
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 11.0.0.dev117+geeca8a4e3.d20221122
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.9.0
xarray           : 2022.11.0
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### LINKED ISSUE #52845: BUG: interchange protocol doesn't work with offsets
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [7]: import pyarrow as pa

In [8]: table = pa.table({"a": pa.array([1,2,3])})

In [9]: from pandas.core.interchange.from_dataframe import from_dataframe

In [10]: from_dataframe(table.slice(1))
Out[10]:
   a
0  2
1  3
2  0
```

### Issue Description

The offset doesn't take effect, `from_dataframe` just returns something of the same length as the original

### Expected Behavior

```python
In [10]: from_dataframe(table.slice(1))
Out[10]:
   a
0  2
1  3
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 478d340667831908b5b4bf09a2787a11a14560c9
python           : 3.8.16.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.0
numpy            : 1.24.2
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : 0.29.33
pytest           : 7.2.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.6.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.6.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.9.0
xarray           : 2023.1.0
xlrd             : None
zstandard        : None
tzdata           : 2022.7
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #49888 (Replace xxxx with the GitHub issue number)
- [x] closes #52845
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This would solve all the "interchange to pandas" issues I'm aware of, so if we could get it in we could really take the interchange protocol for a ride

### CODE REVIEW

File: pandas/core/interchange/from_dataframe.py [hunk 1]
```diff
@@ -406,74 +408,26 @@ def buffer_to_ndarray(
...
     else:
+        data_pointer = ctypes.cast(
+            buffer.ptr + (offset * bit_width // 8), ctypes.POINTER(ctypes_type)
+        )
         return np.ctypeslib.as_array(
             data_pointer, shape=(buffer.bufsize // (bit_width // 8),)
         )
 
 
-def bitmask_to_bool_ndarray(
```
Comment (MarcoGorelli) [inline pandas/core/interchange/from_dataframe.py:423]: looks like this entire function was "dead code" on arrival, https://app.codecov.io/gh/pandas-dev/pandas/blob/main/pandas/core/interchange/from_dataframe.py shows the whole thing as uncovered by tests

Comment (jorisvandenbossche) [reply pandas/core/interchange/from_dataframe.py:423]: Ah, so we actually did have a pure python version of this conversion .. (just not used)

File: pandas/tests/interchange/test_impl.py [hunk 2]
```diff
@@ -104,6 +104,21 @@ def test_large_string_pyarrow():
     assert pa.Table.equals(pa.interchange.from_dataframe(result), table)
 
 
+def test_bitmasks_pyarrow():
+    # GH 52795
+    pa = pytest.importorskip("pyarrow", "11.0.0")
+
+    arr = [3.3, None, 2.1]
+    table = pa.table({"arr": arr})
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply pandas/tests/interchange/test_impl.py]: good one, thanks! that one didn't really work to begin with, so I had to make some slight changes to get it to work

File: pandas/core/interchange/from_dataframe.py [hunk 3]
```diff
@@ -252,7 +258,7 @@ def string_column_to_ndarray(col: Column) -> tuple[np.ndarray, Any]:
         Endianness.NATIVE,
     )
     # Specify zero offset as we don't want to chunk the string data
-    data = buffer_to_ndarray(data_buff, data_dtype, offset=0, length=col.size())
+    data = buffer_to_ndarray(data_buff, data_dtype, offset=0, length=data_buff.bufsize)
```
Comment (MarcoGorelli) [inline pandas/core/interchange/from_dataframe.py:261]: had to change this one for the `large-string` test

File: doc/source/whatsnew/v2.0.1.rst [hunk 4]
```diff
@@ -28,6 +28,8 @@ Bug fixes
 - Bug in :attr:`Series.dt.days` that would overflow ``int32`` number of days (:issue:`52391`)
 - Bug in :class:`arrays.DatetimeArray` constructor returning an incorrect unit when passed a non-nanosecond numpy datetime array (:issue:`52555`)
 - Bug in :func:`Series.median` with :class:`ArrowDtype` returning an approximate median (:issue:`52679`)
+- Bug in :func:`api.interchange.from_dataframe` was returning :class:`DataFrame`'s of incorrect sizes when called on slices (:issue:`52824`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.1.rst:31]: Guess this needs to be moved to 2.0.2 now

**Review by mroeschke — COMMENTED** [review]
One comment otherwise LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MarcoGorelli

**Pull request merged**